### DISCUSSION TITLE
Store fabricJS objects to export in a PDF

### DISCUSSION THREAD

**Discussion by MallauryGerard** [fabricjs/fabric.js#8341] [Q&A]
Hello,

I am using FabricJS as a layer on top of a PDFJS to allow annotating a PDF (page by page).
I am now looking to export the annotated PDF (I have a Laravel/PHP backend).

I see 2 solutions:

**1. Save the fabric objects in the database (JSON). Then recreate each object in the PDF with the FPDF library.**

The problem is that the coordinates of the fabricJS objects do not match the coordinates of FPDF. And so I can't reposition the objects to the right place on the page (nor to the right size).

**2. Save the annotations in base64 in the database (with the toDataUrl function)**

The problem here is the size of the base64 that will have to be saved for each page of the PDF...

Do you have any tips to help me implement solution 1. Or is there a way to convert a json fabric to an image directly in PHP?
Or do you have any other ideas?

Thanks !!

**Comment by ShaMan123** [comment]
1. Checkout http://fabricjs.com/import-pdf
2. https://github.com/joshua-gould/canvas2pdf, should be relatively simple to hack fabric into using this in the `toCanvasElement` method
3. fabric can run on node

**Comment by iocoker** [comment]
@MallauryGerard We implement solution 1, as we are catering for the print industry, which requires every text or shape to remain  vectorized in the PDF.

What we do is, when implementing the `toObject` on the canvas, we go through each of the objects and attach the SVG representation of each object.
`object.toSVG()` will do that for you easily.

To render the PDF, you can use TCPDF since you have PHP. It renders SVGs very easily: [TCPDF](https://tcpdf.org/examples/example_058/)

We have also used [CairoPDF](https://cairosvg.org/) as well as [PDFBox](https://pdfbox.apache.org/)

**Comment by MallauryGerard** [comment]
Thank you both, I will analyse your answers.

**Reply by youssiyouss** [reply]
hello , Did you find after all a good answer for this ?

**Reply by asturur** [reply]
i do some pdf export from fabricJS in this opensource project, you can take a look:
https://github.com/ZaparooProject/zaparoo-designer
